services/unixsocket/service.go

```diff
@@ -0,0 +1,264 @@
+			statBytesReceived:       atomic.LoadInt64(&s.stats.BytesReceived),
+			statPointsParseFail:     atomic.LoadInt64(&s.stats.PointsParseFail),
+			statReadFail:            atomic.LoadInt64(&s.stats.ReadFail),
+			statBatchesTransmitted:  atomic.LoadInt64(&s.stats.BatchesTransmitted),
+			statPointsTransmitted:   atomic.LoadInt64(&s.stats.PointsTransmitted),
+			statBatchesTransmitFail: atomic.LoadInt64(&s.stats.BatchesTransmitFail),
+		},
+	}}
+}
+
+func (s *Service) writer() {
+	defer s.wg.Done()
+
+	for {
+		select {
+		case batch := <-s.batcher.Out():
+			if err := s.PointsWriter.WritePoints(s.config.Database, s.config.RetentionPolicy, models.ConsistencyLevelAny, batch); err == nil {
+				atomic.AddInt64(&s.stats.BatchesTransmitted, 1)
+				atomic.AddInt64(&s.stats.PointsTransmitted, int64(len(batch)))
+			} else {
+				s.Logger.Printf("failed to write point batch to database %q: %s", s.config.Database, err)
+				atomic.AddInt64(&s.stats.BatchesTransmitFail, 1)
+			}
+
+		case <-s.done:
+			return
+		}
+	}
+}
+
+func (s *Service) serve() {
+	defer s.wg.Done()
+
+	buf := make([]byte, MAX_UNIX_SOCKET_PAYLOAD)
+	s.batcher.Start()
+	for {
+
+		select {
+		case <-s.done:
+			// We closed the connection, time to go.
+			return
+		default:
+			// Keep processing.
+			n, _, err := s.conn.ReadFromUnix(buf)
+			if err != nil {
+				if opErr, ok := err.(*net.OpError); ok && !opErr.Temporary() {
+					// Receive singal graceful
+					s.Logger.Printf("unix socket listener closed")
+					continue
+				}
+				atomic.AddInt64(&s.stats.ReadFail, 1)
+				s.Logger.Printf("Failed to read unix socket message: %s", err)
+				continue
+			}
+			atomic.AddInt64(&s.stats.BytesReceived, int64(n))
+
+			bufCopy := make([]byte, n)
+			copy(bufCopy, buf[:n])
+			s.parserChan <- bufCopy
+		}
+	}
+}
+
+func (s *Service) parser() {
+	defer s.wg.Done()
+
+	for {
+		select {
+		case <-s.done:
+			return
+		case buf := <-s.parserChan:
+			points, err := models.ParsePointsWithPrecision(buf, time.Now().UTC(), s.config.Precision)
+			if err != nil {
+				atomic.AddInt64(&s.stats.PointsParseFail, 1)
+				s.Logger.Printf("Failed to parse points: %s", err)
+				continue
+			}
+
+			for _, point := range points {
+				s.batcher.In() <- point
+			}
+			atomic.AddInt64(&s.stats.PointsReceived, int64(len(points)))
+		}
+	}
+}
+
+// Close closes the underlying listener.
+func (s *Service) Close() error {
+	if s.conn == nil {
+		return errors.New("Service already closed")
+	}
+
+	s.conn.Close()
+	s.batcher.Flush()
+	close(s.done)
+	s.wg.Wait()
+
+	// Release all remaining resources.
+	s.done = nil
+	s.conn = nil
+	err := os.Remove(s.config.BindSocket)
+	if err != nil {
+		return err
+	}
+
+	s.Logger.Print("Service closed")
+
+	return nil
+}
+
+// SetLogOutput sets the writer to which all logs are written. It must not be
+// called after Open is called.
+func (s *Service) SetLogOutput(w io.Writer) {
+	s.Logger = log.New(w, "[unixsocket] ", log.LstdFlags)
```

Comment:
nit: while I know other services have used this pattern, it should really be `s.Logger.SetOutput(w)` 😄